mmv1/templates/tgc_next/cai2hcl/resource_converter.go.tmpl

```diff
@@ -0,0 +1,130 @@
+  "github.com/apparentlymart/go-cidr/cidr"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/id"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/structure"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+  "github.com/hashicorp/terraform-plugin-sdk/v2/helper/logging"
+  "google.golang.org/api/bigtableadmin/v2"
+  "google.golang.org/api/googleapi"
+
+  "github.com/GoogleCloudPlatform/terraform-google-conversion/v6/pkg/caiasset"
+  "github.com/GoogleCloudPlatform/terraform-google-conversion/v6/pkg/cai2hcl/converters/utils"
+  "github.com/GoogleCloudPlatform/terraform-google-conversion/v6/pkg/tpgresource"
+  transport_tpg "github.com/GoogleCloudPlatform/terraform-google-conversion/v6/pkg/transport"
+  "github.com/GoogleCloudPlatform/terraform-google-conversion/v6/pkg/verify"
+)
+
+{{- $caiProductBaseUrl := $.CaiProductBaseUrl }}
+{{- $productBackendName := $.CaiProductBackendName $caiProductBaseUrl }}
+{{- $apiVersion := $.CaiApiVersion $productBackendName $caiProductBaseUrl}}
+
+{{if $.CustomCode.Constants -}} 
+    {{- $.CustomTemplate $.CustomCode.Constants true -}}
+{{- end}}
+
+const {{ $.ResourceName -}}AssetType string = "{{ $productBackendName }}.googleapis.com/{{ $.Name -}}"
+const {{ $.ResourceName -}}SchemaName string = "{{ $.TerraformName }}"
+
+type {{ $.ResourceName -}}Converter struct {
+	name   string
+	schema map[string]*schema.Schema
+}
+
+func New{{ $.ResourceName -}}Converter(provider *schema.Provider) models.Converter {
+	schema := provider.ResourcesMap[{{ $.ResourceName -}}SchemaName].Schema
+
+	return &{{ $.ResourceName -}}Converter{
+		name:   {{ $.ResourceName -}}SchemaName,
+		schema: schema,
+	}
+}
+
+// Convert converts asset to HCL resource blocks.
+func (c *{{ $.ResourceName -}}Converter) Convert(asset caiasset.Asset) ([]*models.TerraformResourceBlock, error) {
+	var blocks []*models.TerraformResourceBlock
+	block, err := c.convertResourceData(asset)
+	if err != nil {
+		return nil, err
+	}
+	blocks = append(blocks, block)
+	return blocks, nil
+}
+
+func (c *{{ $.ResourceName -}}Converter) convertResourceData(asset caiasset.Asset) (*models.TerraformResourceBlock, error) {
+	if asset.Resource == nil || asset.Resource.Data == nil {
+		return nil, fmt.Errorf("asset resource data is nil")
+	}
+
+	res := asset.Resource.Data
+	config := utils.NewConfig()
+	d := &schema.ResourceData{}
+
+	hclData := make(map[string]interface{})
+
+{{ range $prop := $.ReadPropertiesForTgc }}
+{{   if $prop.FlattenObject -}}
+// Terraform must set the top level schema field, but since this object contains collapsed properties
```

Comment:
This code is shared with the resource.go template code right? If they are the same, and need to be kept identical, should we move the shared parts to a subtemplate?

Reply:
It shouldn't use `d.Set()` to set values in TGC. Instead, it should set values to `hclData`

I am going to add a TODO comment how to convert FlattenObject